### DISCUSSION TITLE
[Apple] PKCE code_verifier cookie was missing

### DISCUSSION THREAD

**Discussion by marlon360** [nextauthjs/next-auth#6898] [Help] [answered]
I got this error `PKCE code_verifier cookie was missing` when using the Apple provider.
It happens in the callback endpoint.

Does anybody know how to setup the Apple provider correctly?

next: 13.2.1
next-auth: 4.20.1

**Comment by desenmeng** [comment, marked as answer]
add  cookies config
```
  cookies: {
    pkceCodeVerifier: {
      name: "next-auth.pkce.code_verifier",
      options: {
        httpOnly: true,
        sameSite: "none",
        path: "/",
        secure: true,
      },
    },
  },
```

**Reply by marlon360** [reply]
Wow this actually worked! Thank you so much!

**Reply by brucemckayone** [reply]
comming in to the same problem where are you adding this?

**Reply by brucemckayone** [reply]
> comming in to the same problem where are you adding this?

ignore my stupid face it was in the [...nextauth.js] i was in the provider scope

**Reply by michaeldegori** [reply]
I don't know if I'm also stupid, but I'm getting TypeScript errors when trying to add this to the AppleProvider.

```
    AppleProvider({
      clientId: process.env.APPLE_CLIENT_ID!,
      clientSecret: generateAppleSecret(),
      cookies: {
        pkceCodeVerifier: {
          name: 'next-auth.pkce.code_verifier',
          options: {
            httpOnly: true,
            sameSite: 'none',
            path: '/',
            secure: true,
          },
        },
      },
    }),
```

and the TS error is: 
```
Argument of type '{ clientId: string; clientSecret: string; cookies: { pkceCodeVerifier: { name: string; options: { httpOnly: boolean; sameSite: string; path: string; secure: boolean; }; }; }; }' is not assignable to parameter of type 'Omit<OAuthUserConfig<any>, "clientSecret"> & { clientSecret: string; }'.
  Object literal may only specify known properties, and 'cookies' does not exist in type 'Omit<OAuthUserConfig<any>, "clientSecret"> & { clientSecret: string; }'.ts(2345)
```

Please help lol

**Reply by michaeldegori** [reply]
Please also ignore my stupid face haha. Don't add cookies key to the provider, add to the authOptions 🤦🏻‍♂️

**Reply by jeeasistio** [reply]
> add cookies config
> 
> ```
>   cookies: {
>     pkceCodeVerifier: {
>       name: "next-auth.pkce.code_verifier",
>       options: {
>         httpOnly: true,
>         sameSite: "none",
>         path: "/",
>         secure: true,
>       },
>     },
>   },
> ```

why am i still getting error when using google provider?

next: 13.2.4
next-auth: 4.22.0

**Reply by pradeep800** [reply]
hi if you find anything let me know

**Reply by nkeil** [reply]
After you add that cookie config, you must use https:// for your OAuth I think, since google doesn't work on http:// when I have the cookie setup. If you setup a config where you disable the cookie on http://, google should work, but apple won't. When you use https://, enable the cookie and both should work.

**Reply by Suprefuner** [reply]
Hi I added the cookie config to Google provider but still getting 'PKCE code_verifier cookie was missing' error
anyone could help me out? Here is my code
next: 13.2.4
next-auth: 4.22.1

```
import NextAuth from "next-auth"
import GoogleProvider from "next-auth/providers/google"
import { PrismaAdapter } from "@next-auth/prisma-adapter"
import prisma from "../../../../prisma/client"

const handler = NextAuth({
  adapter: PrismaAdapter(prisma),
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
  ],
  cookies: {
    pkceCodeVerifier: {
      name: "next-auth.pkce.code_verifier",
      options: {
        httpOnly: true,
        sameSite: "none",
        path: "/",
        secure: true,
      },
    },
  },
})
export { handler as GET, handler as POST }
```

**Reply by nkeil** [reply]
@Suprefuner Are you running on http or https? The `secure: true` in your cookie config will only work on https. If you want http try setting that to false when not in production.

**Reply by valentepablo** [reply]
> Hi I added the cookie config to Google provider but still getting 'PKCE code_verifier cookie was missing' error anyone could help me out? Here is my code next: 13.2.4 next-auth: 4.22.1
> 
> ```
> import NextAuth from "next-auth"
> import GoogleProvider from "next-auth/providers/google"
> import { PrismaAdapter } from "@next-auth/prisma-adapter"
> import prisma from "../../../../prisma/client"
> 
> const handler = NextAuth({
>   adapter: PrismaAdapter(prisma),
>   // Configure one or more authentication providers
>   providers: [
>     GoogleProvider({
>       clientId: process.env.GOOGLE_CLIENT_ID,
>       clientSecret: process.env.GOOGLE_CLIENT_SECRET,
>     }),
>   ],
>   cookies: {
>     pkceCodeVerifier: {
>       name: "next-auth.pkce.code_verifier",
>       options: {
>         httpOnly: true,
>         sameSite: "none",
>         path: "/",
>         secure: true,
>       },
>     },
>   },
> })
> export { handler as GET, handler as POST }
> ```

I'm having the same issue running the same versions of Next and Next-auth. Could you find a solution for this?

**Reply by Suprefuner** [reply]
> @Suprefuner Are you running on http or https? The `secure: true` in your cookie config will only work on https. If you want http try setting that to false when not in production.

@nkeil Hi, I'm running on http since just in development stage, but I have tried adding or removing cookies config still doesn't work. You guys tried and everything just work just fine?

**Reply by Suprefuner** [reply]
> > Hi I added the cookie config to Google provider but still getting 'PKCE code_verifier cookie was missing' error anyone could help me out? Here is my code next: 13.2.4 next-auth: 4.22.1
> > ```
> > import NextAuth from "next-auth"
> > import GoogleProvider from "next-auth/providers/google"
> > import { PrismaAdapter } from "@next-auth/prisma-adapter"
> > import prisma from "../../../../prisma/client"
> > 
> > const handler = NextAuth({
> >   adapter: PrismaAdapter(prisma),
> >   // Configure one or more authentication providers
> >   providers: [
> >     GoogleProvider({
> >       clientId: process.env.GOOGLE_CLIENT_ID,
> >       clientSecret: process.env.GOOGLE_CLIENT_SECRET,
> >     }),
> >   ],
> >   cookies: {
> >     pkceCodeVerifier: {
> >       name: "next-auth.pkce.code_verifier",
> >       options: {
> >         httpOnly: true,
> >         sameSite: "none",
> >         path: "/",
> >         secure: true,
> >       },
> >     },
> >   },
> > })
> > export { handler as GET, handler as POST }
> > ```
> 
> I'm having the same issue running the same versions of Next and Next-auth. Could you find a solution for this?

@valentepablo Hi, still stuck on this, trying to get help from others, will update if I found the solution!

**Reply by nkeil** [reply]
> > @Suprefuner Are you running on http or https? The `secure: true` in your cookie config will only work on https. If you want http try setting that to false when not in production.
> 
> @nkeil Hi, I'm running on http since just in development stage, but I have tried adding or removing cookies config still doesn't work. You guys tried and everything just work just fine?

See if this works:
```ts
cookies: {
  pkceCodeVerifier: {
    name: 'next-auth.pkce.code_verifier',
    options: {
      httpOnly: true,
      sameSite: 'none',
      path: '/',
      secure: process.env.NODE_ENV === 'production',
    },
  },
},
```

**Reply by Suprefuner** [reply]
> ```ts
> pkceCodeVerifier: {
>     name: 'next-auth.pkce.code_verifier',
>     options: {
>       httpOnly: true,
>       sameSite: 'none',
>       path: '/',
>       secure: process.env.NODE_ENV === 'production',
>     },
>   },
> ```

@nkeil Hi, thanks for your reply, but still doesn't work. actually my next-auth behave weird not sure if it's my problem or what.
I'm working on a personal project and facing 4 weird situations
1. no matter I input the wrong email or password for the credential login, next-auth response ok: true and status:200 , doesn't affect the feature just feel a bit confuse or I login successfully next-auth still response with the error:OAuthcallback 
2. for the github login, I have to login and logout with credential first then I could create with github successfully otherwise will get an `state cookie missing` or `state mismatch` error
3. for the github login, even I create a user in my DB(I use mongoDB) successfully but I can't get the session using getServerSession which means I can't login with github account
4. google login, no luck to success, keep getting `PKCE code_verifier cookie was missing` error

do you mind if I share you the code , if you have time help me to take a look?

**Reply by nkeil** [reply]
@Suprefuner Sure, I can take a look

**Reply by Lenoxel** [reply]
> add cookies config
> 
> ```
>   cookies: {
>     pkceCodeVerifier: {
>       name: "next-auth.pkce.code_verifier",
>       options: {
>         httpOnly: true,
>         sameSite: "none",
>         path: "/",
>         secure: true,
>       },
>     },
>   },
> ```
@desenmeng wow, fantastic!

I was running into this issue has some weeks, and now it's solved here in the Apple Provider.

Thank you!

**Reply by trevorpfiz** [reply]
> 4. google login, no luck to success, keep getting `PKCE code_verifier cookie was missing` error

I am getting this error in development with the Discord and Google providers when I set:

```
pkceCodeVerifier: {
    name: 'next-auth.pkce.code_verifier',
    options: {
      httpOnly: true,
      sameSite: 'none',
      path: '/',
      secure: process.env.NODE_ENV === 'production',
    },
  },
```

any way to fix this?

**Reply by jhirvioja** [reply]
Why does this work, and does it introduce any security flaws? I see it introduces `sameSite: 'none'` to the options, which I am hesitant to use..

**Reply by matt-wigg** [reply]
I get the same error with the AzureADProvider [/discussions/10502](https://github.com/nextauthjs/next-auth/discussions/10502)

**Reply by KubaVejrazka** [reply]
> > ```ts
> > pkceCodeVerifier: {
> >     name: 'next-auth.pkce.code_verifier',
> >     options: {
> >       httpOnly: true,
> >       sameSite: 'none',
> >       path: '/',
> >       secure: process.env.NODE_ENV === 'production',
> >     },
> >   },
> > ```
> 
> @nkeil Hi, thanks for your reply, but still doesn't work. actually my next-auth behave weird not sure if it's my problem or what. I'm working on a personal project and facing 4 weird situations
> 
> 1. no matter I input the wrong email or password for the credential login, next-auth response ok: true and status:200 , doesn't affect the feature just feel a bit confuse or I login successfully next-auth still response with the error:OAuthcallback
> 2. for the github login, I have to login and logout with credential first then I could create with github successfully otherwise will get an `state cookie missing` or `state mismatch` error
> 3. for the github login, even I create a user in my DB(I use mongoDB) successfully but I can't get the session using getServerSession which means I can't login with github account
> 4. google login, no luck to success, keep getting `PKCE code_verifier cookie was missing` error
> 
> do you mind if I share you the code , if you have time help me to take a look?

Hey, have you found a solution? I need to implement google oauth and I have the same problem. No matter how i set up my cookies, i still get the "InvalidCheck: PKCE code_verifier cookie was missing." error.

Thanks!

**Reply by KubaVejrazka** [reply]
Hey everyone,

I've found the cause of this problem (in my case) and somewhat of a solution.

I was hosting the website to a public ip by port forwarding my localhost, which caused mismatch and led to this error.

Once I set the redirect proxy to `<public ip>/auth`, I finally got rid of the error, however I still wasn't able to complete the callback because SvelteKit (which I'm building my web app with) kept adding the localhost port to my public URI. But the oAuth request went through without any problems.

The way I solved all of these issues was by hosting my app directly to a public IP. You can do this for free by using a service like Vercel. In my auth config I removed both the code verifier cookie config recommended by this issue's answer and the redirect proxy, and I set trustHost to true.

Conclusion:
If you are port forwarding localhost, try setting a redirect proxy in your auth config to `<public ip>/<auth route>`.
If you still face issues (for example the localhost port being added to your callback URI) that you don't know how to fix, try hosting your app directly to a public ip.

I hope I helped someone :)

**Reply by pranavmappoli** [reply]
> ```ts
> cookies: {
>   pkceCodeVerifier: {
>     name: 'next-auth.pkce.code_verifier',
>     options: {
>       httpOnly: true,
>       sameSite: 'none',
>       path: '/',
>       secure: process.env.NODE_ENV === 'production',
>     },
>   },
> },
> ```

@trevorpfiz Did you find any solution ?

**Comment by 25sahilsingh** [comment]
so i don't know much but i was having the same issue ....but my problem has something to do with link in which i was accessing the web ...for example in GitHub i have given my network link somethink like http://192.168.X.X:3000. but i was trying to access it from localhost:3000
